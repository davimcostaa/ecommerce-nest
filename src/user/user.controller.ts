import { Body, Controller, Get, Post, UsePipes, ValidationPipe, Param, Patch } from '@nestjs/common';
import { Roles } from 'src/decorators/roles.decorator';
import { UserId } from 'src/decorators/user-id.decorator';
import { CreateUserDTO } from './dtos/createUser.dto';
import { ReturnUserDTO } from './dtos/returnUser.dto';
import { UpdatePasswordDTO } from './dtos/update-password.dto';
import { UserEntity } from './entities/user.entity';
import { UserType } from './enum/userType.enum';
import { UserService } from './user.service';

@Controller('user')
export class UserController {

    constructor(private readonly userService: UserService) {}

    @UsePipes(ValidationPipe)
    @Post()
    async createUser(@Body() createUser: CreateUserDTO): Promise<UserEntity> {
        return this.userService.createUser(createUser);
    }

    @Roles(UserType.Admin)
    @Get()
    async getAllUser():Promise<ReturnUserDTO[]> {
        return (await this.userService.getAllUsers()).map(
            (userEntity) => new ReturnUserDTO(userEntity))
    }

    @Roles(UserType.Admin)
    @Get('/:userId')
    async getUserById(@Param('userId') userId: number): Promise<ReturnUserDTO> {
        return new ReturnUserDTO(await this.userService.getUserByIdUsingRelations(userId));
    }

    @Roles(UserType.Admin, UserType.User)
    @Patch()
    async updateUserPassword(
        @UserId() userId: number,
        @Body() updateUserPassword: UpdatePasswordDTO
    ):Promise<UserEntity> {
        return this.userService.updatePasswordUser(updateUserPassword, userId)
    }

}
