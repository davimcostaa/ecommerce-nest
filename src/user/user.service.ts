import { BadRequestException, Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { hash } from 'bcrypt';
import { Repository } from 'typeorm';
import { CreateUserDTO } from './dtos/createUser.dto';
import { UserEntity } from './entities/user.entity';

@Injectable()
export class UserService {

    constructor (
        @InjectRepository(UserEntity)
        private readonly userRepository: Repository<UserEntity>
    ) {}

    async createUser(createUserDTO: CreateUserDTO): Promise<UserEntity> {

        const user = await this.findUserByEmail(createUserDTO.email).catch(() => undefined)

        if(user) {
            throw new BadRequestException('email registered in system')
        }

        const saltOrRounds = 10;
        const passwordHashed = await hash(createUserDTO.password, saltOrRounds);

        return this.userRepository.save({
            ...createUserDTO,
            typeUser: 1,
            password: passwordHashed
        })
    }

    async getUserByIdUsingRelations(userId: number): Promise<UserEntity> {
        
        const user = await this.userRepository.findOne({
            where: {
                id: userId
            },
            relations: {
                addresses: {
                    city: {
                        state: true
                    }
                }
            }
        })

        if (!user) {
            throw new NotFoundException
        }

        return user
    }

    async getAllUsers(): Promise<UserEntity[]> {
        return this.userRepository.find()
    } 

    async findUserById(userId: number): Promise<UserEntity> {
        const user = await this.userRepository.findOne({
            where: {
                id: userId
            }
        })

        if (!user) {
            throw new NotFoundException('UserId not found!')
        }

        return user
    }

    async findUserByEmail(email: string): Promise<UserEntity> {
        const user = await this.userRepository.findOne({
            where: {
                email: email
            }
        })

        if (!user) {
            throw new NotFoundException('Email not found!')
        }

        return user
    }
 }
