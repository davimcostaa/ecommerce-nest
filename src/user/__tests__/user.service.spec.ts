import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { UserEntity } from '../entities/user.entity';
import { UserService } from '../user.service';
import { createUserMock } from '../__mocks__/createUser.mock';
import {  userEntityMock } from '../__mocks__/user.mock';

describe('UserService', () => {
  let service: UserService;

  let userRepository: Repository<UserEntity>

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [UserService,
        {
          provide: getRepositoryToken(UserEntity),
          useValue: {
            findOne: jest.fn().mockResolvedValue(userEntityMock),
            save: jest.fn().mockResolvedValue(userEntityMock)
          } 
        },
      ],
    }).compile();

    service = module.get<UserService>(UserService);
    userRepository = module.get<Repository<UserEntity>>(getRepositoryToken(UserEntity))
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
    expect(userRepository).toBeDefined();
  });

  it('should return user in findUserByEmail', async () => {
    const user = await service.findUserByEmail(userEntityMock.email);

    expect(user).toEqual(userEntityMock)
  });

  it('should return error in findUserByEmail', async () => {
    jest.spyOn(userRepository, 'findOne').mockResolvedValue(null)

    expect(service.findUserByEmail(userEntityMock.email)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should return error in findUserByEmail (error request)', async () => {
    jest.spyOn(userRepository, 'findOne').mockRejectedValueOnce(new Error())

    expect(service.findUserByEmail(userEntityMock.email)).rejects.toThrowErrorMatchingSnapshot();
  });

  
  it('should return user in findUserById', async () => {
    const user = await service.findUserById(userEntityMock.id);

    expect(user).toEqual(userEntityMock)
  });

  it('should return error in findUserById', async () => {
    jest.spyOn(userRepository, 'findOne').mockResolvedValue(null)

    expect(service.findUserById(userEntityMock.id)).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should return user in getUserByIdUsingRelations', async () => {
    const user = await service.getUserByIdUsingRelations(userEntityMock.id);

    expect(user).toEqual(userEntityMock)
  });

  it('should return error if user exists', async () => {
    expect(service.createUser(createUserMock)).rejects.toThrowErrorMatchingSnapshot()
  });

  it('should return user if user not exists', async () => {
    jest.spyOn(userRepository, 'findOne').mockResolvedValue(null);

    const user = await service.createUser(createUserMock);

    expect(user).toEqual(userEntityMock)

  });

});
