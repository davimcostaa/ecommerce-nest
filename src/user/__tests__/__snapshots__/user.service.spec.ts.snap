// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserService should return error if user exists 1`] = `"email registered in system"`;

exports[`UserService should return error in findUserByEmail (error request) 1`] = `""`;

exports[`UserService should return error in findUserByEmail 1`] = `"Email not found!"`;

exports[`UserService should return error in findUserById 1`] = `"UserId not found!"`;
